FBC Documentation (#193)
redhat-appstudio/docs.appstudio.io: docs/modules/ROOT/pages/how-to-guides/proc_maintaining_operator_graphs.adoc

:_content-type: PROCEDURE
:myfunctionone: maintaining_operator_graphs

[id="maintaining_operator_graphs_{context}"]
= Maintaining operator graphs using the information on the File-based catalogs

> File-based catalogs (FBC) provide a streamlined approach to storing and distributing metadata about operators written for the Operator Lifecycle Manager (OLM) and can be built and distributed uisng {ProductName}. You can use {ProductName} to build, test, and release multiple components from a single FBC repo. Here release means that {ProductName} releases an FBC image to the Red Hat’s public index.

Reviewer: > Here release means that {ProductName} releases an FBC image to the Red Hat’s public index.

I think we can leave this part out. This isn't necessarily true as it depends on what managed workspace is being targeted and what the actions are that it is configured to take.

> File-based catalogs simplify the management of graph updates, offering enhanced control and ownership of operator graphs. With a choice of JSON or YAML formats, File-based catalogs ensure backward compatibility while optimizing efficiency.

Reviewer: The language implies that you are comparing against something but you never mention what comes before.

* *Convenient management and flexibility:* Compared to SQLite, FBC offers a more convenient solution. Rather than adding one set of graph updates at a time, users can define all updates in one place using File-based catalogs. This centralized management capability simplifies the process and improves overall efficiency.

* *Improved ownership and analysis:* FBC provides users with improved ownership and control over their graphs. They offer a clean and declarative template that facilitates comprehensive analysis and sanity checks. Users can easily determine upgrade possibilities and recover from error states, no longer relying on limited configuration options found in localized CSV files.

For additional information, see link:https://olm.operatorframework.io/docs/reference/file-based-catalogs/[File based Catalog].

== Setting up an FBC Compatible git repository

An FBC-compatible git repository has:

> * A Docker file at the root of the FBC repository. This file has the `ose-operator-registry` binary images. 

Reviewer: > This file has the `ose-operator-registry` binary images. 
 
I think that this language isn't clear enough. The registry.redhat.io/openshift4/ose-operator-registry image needs to be the parent image for the Dockerfile.

> * A Docker file at the root of the FBC repository. This file has the `ose-operator-registry` binary images. 

Reviewer: Not just _any_ Dockerfile, but a Dockerfile corresponding the `uri` as indicated in the Devfile. Additionally, each Dockerfile should have a tag corresponding to the target OCP version. I think we should move this part to after the devfile due to its dependence on it.

Author: I've tried a new flow here. Please let me know your thoughts. If you do not like it, I'll move the content as you suggested.

Reviewer: That flow feels better, thanks!

> A `yaml` file (for example, `catalog-config.yml`) that targets different Red Hat OpenShift Container Platform (OCP) versions. This `yaml` file contains important metadata of the operator. 

Reviewer: This feels slightly incorrect. There should be a `catalog.yml` for each FBC image that is produced (each FBC component targets one OCP version based on the parent reference used).

Also, I think that this file name needs to be consistent with the Dockerfile and its location needs to be consistent with the `buildContext` in the devfile. Therefore, should we move this part to after the devfile is introduced?

    builders:
      - olm.builder.basic
  - name: "4.12"
    destination:
      baseImage: registry.redhat.io/openshift4/ose-operator-registry:v4.12 This template can be hosted
      workingDir: .
    builders:
      - olm.builder.basic
----
<1> The OCP version.
<2> The binary image that the system uses to build the FBC fragments.

* For every targeted OCP version, you should have a directory with a `devfile` that {ProductName} uses to detect the type of component. It also contains information about how to set up a build. 

+	
*Example:*

+
[source,yaml]
----
schemaVersion: 2.2.0
metadata:
  name: fbc-4.12
  displayName: FBC 4.12
  description: 'File based catalog'
  language: fbc <1>
  provider: Red Hat
components:
  - name: image-build
    image:
      imageName: ""
      dockerfile:
        uri: catalog.Dockerfile
        buildContext: "4.12"
  - name: kubernetes
    kubernetes:
      inlined: placeholder
    attributes:
      deployment/container-port: 50051
      deployment/cpuRequest: "100m"
      deployment/memoryRequest: 512Mi
      deployment/replicas: 1
      deployment/storageRequest: "0"
commands:
  - id: build-image
    apply:
      component: image-build"
----

> <1>  Indicates the component that {ProductName} builds.

Reviewer: Should this be a unique reference with the prior example?

Maybe 
> Indicates the component type that [...]

instead

== Adding a component
Users should manage all the FBC components through a single application. This centralized management enables synchronized updates for graphs across all OpenShift versions.

.Prerequisites

* You have access to link:https://console.redhat.com/beta/hac/application-pipeline[App Studio].

> * You have FBC compatible git repository.

Reviewer: Compatibility doesn't feel like the right word. FBC consistent? A git repository with FBC components?

* You have to link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/Import-code/proc_importing_code/#creating-a-red-hat-container-registry-token[create a Red Hat Container Registry token].

.Procedures

. Go to App Studio and select Create application.
. In the *Git repository URL* field, enter the URL of your FBC-compatible Git repository.

> . Select *Import code*. The system displays the Configure your components for deployment page.

Reviewer: Should we prompt the user to deselect all components which are _not_ FBC or that they don't want to onboard? Or to ensure that all desired components are present?

Reviewer 2: It is possible to import the same repo multiple times. RHTAP just changes the automatic name, it doesn't de-select that component. The UI can be used to de-select detected components, and as long as one is selected, they can import.

Author: Thank you, Andrew and Matt
As per the conversation, I think we do not need any updates here.

. Disable the Default build pipeline.
. In the Secrets section, select *Add secret*.
. From the *Secret type* dropdown list, select *Image pull secret*.
. In the *Select or enter name*, enter *registry-redhat-io-docker*. 

> . In the *Value* field, copy paste or upload the downloaded secret.

Reviewer: Does this belong here? Shouldn't we just pop out to somewhere else in the documentation where we show users how to add secrets for registries? The reason we need to do it here is because of the requirement we impose for a registry.redhat.io parent image.

Author: I've mentioned that in the prerequisites. We don't have the steps mentioned above on the linked page, hence I added the steps.

![image](https://github.com/redhat-appstudio/docs.appstudio.io/assets/90042568/963abf39-ece4-4e8f-8d2b-c54005f82b07)

Reviewer: Thanks, you can resolve this thread.

:_content-type: PROCEDURE
:myfunctionone: maintaining_operator_graphs

[id="maintaining_operator_graphs_{context}"]

> = Maintaining operator graphs using the information on the File-based catalogs

Reviewer 2: Could the title be simplified? Would `maintaining operator graphs using file-based catalogs` be as good? Does it need to call out FBC, or could that be the mechanism that RHTAP uses and describes in the doc? Is "maintaining an operator graph" meaningful to all operator developers?

> File-based catalogs (FBC) provide a streamlined approach to store and distribute metadata about operators written for the Operator Lifecycle Manager (OLM). You can use {ProductName} to build, test, and release multiple components from a single FBC repo. 

Reviewer 2: Maybe I'm looking at this wrong, but I don't see this paragraph displayed when viewing the preview link. Underneath the title is "FBCs represent...", so I don't see the acronym spelled out before it gets used.

> File-based catalogs (FBC) provide a streamlined approach to store and distribute metadata about operators written for the Operator Lifecycle Manager (OLM). You can use {ProductName} to build, test, and release multiple components from a single FBC repo. 

Reviewer: This documentation has an "interesting audience" in that it is _currently_ largely intended for internal users who are familiar with the SQLite release process.

We can probably simplify the wording in the document to use "Operator catalog"/"catalog" convention for our own sakes as long as we clearly call out the change from SQLite -- I think this might be all we currently do?

> * *Convenient management and flexibility:* Compared to SQLite, FBC offers a more convenient solution for managing graph updates. Users can define all updates in one place using file-based catalogs, streamlining the process and improving overall efficiency. This centralized management capability simplifies the process and improves overall efficiency.

Since we use `fbc` as abbreviations internally, we should still introduce that acronym, but we can change most usages to a generic "catalog" term.

> FBCs represent a departure from the previous method of defining operator graphs using a CSV file. Previously, operator developers defined operator graphs implicitly in CSV files. This made made it difficult to maintain and update the graphs, as any changes required updating the CSV files and restarting the OLM controller.

Reviewer 2: typo here - This made made it...

Reviewer 2: ```suggestion
FBCs represent a departure from the previous method of defining operator graphs using a CSV file. Previously, operator developers defined operator graphs implicitly in CSV files. This made it difficult to maintain and update the graphs, as any changes required updating the CSV files and restarting the OLM controller.
```

> FBCs simplyfy graph management, offering enhanced control and ownership of operator graphs. With a choice of JSON or YAML formats, File-based catalogs ensure backward compatibility while optimizing efficiency.

Reviewer 2: ```suggestion
FBCs simplify graph management, offering enhanced control and ownership of operator graphs. With a choice of JSON or YAML formats, File-based catalogs ensure backward compatibility while optimizing efficiency.
```

The advantages it offers are:

* *Enhance control and simplify graph updates:* With FBC, graph updates are consolidated into a single catalog file. This eliminates the need for iterative updates and simplifies the management of complex graphs. Users can easily manipulate and control their graphs by working with a centralized catalog file.

* *Convenient management and flexibility:* Compared to SQLite, FBC offers a more convenient solution. Rather than adding one set of graph updates at a time, users can define all updates in one place using File-based catalogs. This centralized management capability simplifies the process and improves overall efficiency.

> * *Improved ownership and analysis:* FBC provides users with improved ownership and control over their graphs. They offer a clean and declarative template that facilitates comprehensive analysis and surface level checks. Users can easily determine upgrade possibilities and recover from error states, no longer relying on limited configuration options found in localized CSV files.

Reviewer 2: If the first bullet promises "Enhance control and simplify graph updates", should the third bullet after promising "Improved ownership and analysis", go on to say "FBC provides users with improved ownership and control over their graphs"?
They both mention control over the graph.

Author: Yes, they convey similar information, but they emphasize slightly different aspects of the benefits of using FBC. I've updated the content to remove any overlapping information.

> * *Improved ownership and analysis:* FBC provides users with improved ownership and control over their graphs. They offer a clean and declarative template that facilitates comprehensive analysis and sanity checks. Users can easily determine upgrade possibilities and recover from error states, no longer relying on limited configuration options found in localized CSV files.

Reviewer 2: I thought the style guide recommended surface level checks instead of sanity? Should we say something different there?

For additional information, see link:https://olm.operatorframework.io/docs/reference/file-based-catalogs/[File based Catalog].

== Setting up an FBC Compatible git repository

An FBC Compatible Git repository has the following directory structure:

----
Project Root
│

> ├─ Dockerfile (for example, catalog.Dockerfile)

Reviewer: If you are showing an example of multiple context directories, you should have an example of multiple Dockerfiles as these should be 1:1.

Author: Removed the multiple context directories.

│

> ├─ Directory for the targetted OCP version (for example, 4.12)

Reviewer: s/targetted/targeted/g

+	
*Example:*

+
[source,catalog.Dockerfile]
----
# The base image is expected to contain
# /bin/opm (with a serve subcommand) and /bin/grpc_health_probe
FROM registry.redhat.io/openshift4/ose-operator-registry:v4.12

# Configure the entrypoint and command
ENTRYPOINT ["/bin/opm"]
CMD ["serve", "/configs", "--cache-dir=/tmp/cache"]

# Copy declarative config root into image at /configs and pre-populate serve cache
ADD catalog /configs
RUN ["/bin/opm", "serve", "/configs", "--cache-dir=/tmp/cache", "--cache-only"]

# Set DC-specific label for the location of the DC root directory
# in the image
LABEL operators.operatorframework.io.index.configs.v1=/configs
----

* *`catalog.yaml`*: This file contains important metadata about the operator. There should be a `catalog.yml` file for each FBC image that is produced.

+
*Example:*

+
[source,yaml]
----
schema: olm.composite.catalogs
catalogs:
  - name: "4.13" <1>
    destination:
      baseImage: registry.redhat.io/openshift4/ose-operator-registry:v4.13 <2>
      workingDir: .
    builders:
      - olm.builder.basic
  - name: "4.12"
    destination:

>       baseImage: registry.redhat.io/openshift4/ose-operator-registry:v4.12 This template can be hosted

Reviewer 2: Is "this template can be hosted" an annotation, or part of the example file? It feels out of place as currently styled, but I could be misunderstanding the purpose.

> * *`devfile.yaml`*: This file is used by {ProductName} to detect the type of component and to set up a build.

Reviewer 2: Should we mention that the language: fbc field must be set here in order to properly build it? I could see it being missed by only calling it out in the annotations below.

----
<1> The OCP version.
<2> The parent image that the system uses to build the FBC fragments.

* *`devfile.yaml`*: This file is used by {ProductName} to detect the type of component and to set up a build.

+	
*Example:*

+
[source,yaml]
----
schemaVersion: 2.2.0
metadata:
  name: fbc-4.12
  displayName: FBC 4.12
  description: 'File based catalog'
  language: fbc <1>
  provider: Red Hat
components:
  - name: image-build
    image:
      imageName: ""
      dockerfile:
        uri: catalog.Dockerfile <2>
        buildContext: "4.12" <3>
  - name: kubernetes
    kubernetes:
      inlined: placeholder
    attributes:
      deployment/container-port: 50051
      deployment/cpuRequest: "100m"
      deployment/memoryRequest: 512Mi
      deployment/replicas: 1
      deployment/storageRequest: "0"
commands:
  - id: build-image
    apply:
      component: image-build"
----

> <1> Indicates the component type that {ProductName} builds.

Reviewer: I think we can be slightly stronger for this statement. This informs {ProductName} that the component being build is a FBC component. Auto-detection will not work to identify FBC, so this is required to inform the system of the type.

<2> Indicates the Dockerfile as defined earlier.

> <3> Indicates the tag corresponding to the targetted OCP version as defined earlier.

Reviewer: These still confuse me being the same numbers as the previous examples. Is this reuse the norm?

Reviewer: that's fine as it is.

> <3> Indicates the tag corresponding to the targetted OCP version as defined earlier.

Reviewer: <3> does not indicate the tag -- the tag corresponding to the OCP version is represented in the Dockerfile (i.e. the FROM statement). This is the context directory needed to specify the FBC component corresponding to the OCP version being targeted (i.e. according to the referenced Dockerfile).

== Adding a component

> Users should manage all the FBC components through a single application. This centralized management enables synchronized updates for graphs across all OpenShift versions.

Reviewer 2: Is there more context we should include here? Do we need all of their FBC components across all of their workspaces in a single FBC application? All FBC components in their workspace in a single FBC app in that workspace? What's the intended scope here?

.Procedures

. Go to App Studio and select Create application.
. In the *Git repository URL* field, enter the URL of your FBC-compatible Git repository.

> . Select *Import code*. The system displays the Configure your components for deployment page.

Author: I'm not sure if we bold page names. Do we?

. Disable the Default build pipeline.

> . In the Secrets section, select *Add secret*.

Author: I'm not sure if we bold section names. Do we?

. From the *Secret type* dropdown list, select *Image pull secret*.
. In the *Select or enter name*, enter *registry-redhat-io-docker*. 

> . In the *Value* field, copy paste or upload the downloaded secret.

Reviewer 3: ```suggestion
. In the *Value* field, copy and paste or upload the downloaded secret.
```

. Select *Create*.

> . Select *Create application*. The system displays a Manage build pipelines pop-up and tries to send a pull request.

Reviewer 3: ```suggestion
. Select *Create application*. The system displays a *Manage build pipelines* pop-up and tries to send a pull request.
```

Author: I'm not sure if we bold page names. Do we?

. In the Secrets section, select *Add secret*.
. From the *Secret type* dropdown list, select *Image pull secret*.
. In the *Select or enter name*, enter *registry-redhat-io-docker*. 
. In the *Value* field, copy paste or upload the downloaded secret.
. Select *Create*.

> . Select *Create application*. The system displays a Manage build pipelines pop-up and sends a pull request.

Reviewer 2: For this to succeed, they will have to grant access or have already granted access to the repo they're trying to add (for our GH app to be able to access it).

> . If you have not already installed the App Studio GitHub application, select Install GitHub application on the “Manage build pipelines” popup page, select *Install Github application*. 

Reviewer 3: ```suggestion
. If you have not already installed the App Studio GitHub application, select *Install GitHub application* on the *Manage build pipelines* popup page, select *Install Github application*. 
```
Although this also indicates clicking "Install GitHub application" somewhere twice. Not sure if that's intentional/accurate or a dupe.

> NOTE: If you want to restrict the GitHub application’s access to certain repositories only, use the Only select repositories option in GitHub during the installation.

Reviewer 3: ?
```suggestion
NOTE: If you want to restrict the GitHub application’s access to certain repositories only, use the *Only select repositories* option in GitHub during the installation.
```

> . Returning to the “Manage build pipelines” page in {ProductName}, select *Send pull request*.

Reviewer 3: I'm not really sure what the consistency is supposed to be across page titles vs GUI labels and buttons and whatnot, but just pointing out where there seems to be possible inconsistencies:
```suggestion
. Returning to the *Manage build pipelines* page in {ProductName}, select *Send pull request*.
```

. Select *Merge in GitHub*.

> . In GitHub, merge the pull request from the “red-hat-trusted-app-pipeline” bot.

Reviewer 3: Suggest backticks or asterisks for a name like this:
```suggestion
. In GitHub, merge the pull request from the `red-hat-trusted-app-pipeline` bot.
```

. Let {ProductName} complete PipelineRun for the upgraded build pipeline.

.Verification

. Go to *Activity > Pipeline* runs.

. Select the most recent *PipelineRun*.

> . View the +++ <span class="popup"><code>fbc-label-check</code><span class="popuptext">Ensures that FBC image have all necessary labels.</span> </span>, <span class="popup"><code>fbc-validate</code><span class="popuptext">Checks the existence and functionality of binaries, and runs validation on configs within the image.</span> </span>, and the <code>fbc-related-image-check</code> task within the build pipeline. Ensure that the build process is executing these tasks and not indicating a “skip checks” status. If these tasks show a green checkmark, it indicates that the build process has validated the fbc components against an existing link:https://enterprisecontract.dev/docs/ec-policies/release_policy.html#labels_package[Enterprise Contract] policy. However, if they do not show a green checkmark, it indicates some potential issues in FBC components. +++

Reviewer 2: I could be wrong, but doesn't the fbc-label-check task have its own logic? and the contract that we check for it come release time ensures that the was task was run during the build?

Author: @arewm WDYT?

Reviewer: fbc-label-check is now present as part of EC and is no longer a test at build time.

> . View the +++ <span class="popup"><code>fbc-label-check</code><span class="popuptext">Ensures that FBC image have all necessary labels.</span> </span>, <span class="popup"><code>fbc-validate</code><span class="popuptext">Checks the existence and functionality of binaries, and runs validation on configs within the image.</span> </span>, and the <code>fbc-related-image-check</code> task within the build pipeline. Ensure that the build process is executing these tasks and not indicating a “skip checks” status. If these tasks show a green checkmark, it indicates that the build process has validated the fbc components against an existing link:https://enterprisecontract.dev/docs/ec-policies/release_policy.html#labels_package[Enterprise Contract] policy. However, if they do not show a green checkmark, it indicates some potential issues in FBC components. +++

Reviewer 2: I don't think the enterprise contract link here is correctly configured. It didn't show properly for me in the preview link at least.

> . View the +++ <span class="popup"><code>fbc-label-check</code><span class="popuptext">Ensures that FBC image have all necessary labels.</span> </span>, <span class="popup"><code>fbc-validate</code><span class="popuptext">Checks the existence and functionality of binaries, and runs validation on configs within the image.</span> </span>, and the <code>fbc-related-image-check</code> task within the build pipeline. Ensure that the build process is executing these tasks and not indicating a “skip checks” status. If these tasks show a green checkmark, it indicates that the build process has validated the fbc components against an existing link:https://enterprisecontract.dev/docs/ec-policies/release_policy.html#labels_package[Enterprise Contract] policy. However, if they do not show a green checkmark, it indicates some potential issues in FBC components. +++

Reviewer 3: A few wording/style nits (also suggest breaking up the step a little for readability), but also because the `+++` is at the end of the paragraph, it's making the `link:` come through verbatim. It seems like you're done with HTML tags after the `fbc-related-image-check` mention, so maybe you can move the `+++` up closer to that so that the `link:` can render correctly?
```suggestion
. View the +++ <span class="popup"><code>fbc-label-check</code><span class="popuptext">Ensures that FBC images have all necessary labels.</span> </span>, <span class="popup"><code>fbc-validate</code><span class="popuptext">Checks the existence and functionality of binaries, and runs validation on configs within the image.</span> </span>, and the <code>fbc-related-image-check</code> +++ task within the build pipeline. Ensure that the build process is executing these tasks and not indicating a “skip checks” status.
+
If these tasks show a green checkmark, it indicates that the build process has validated the FBC components against an existing link:https://enterprisecontract.dev/docs/ec-policies/release_policy.html#labels_package[Enterprise Contract] policy. However, if they do not show a green checkmark, it indicates some potential issues in FBC components.
```

> NOTE: We now have the image, next step is to release the image for which we need to set up the workspace.

Reviewer 3: Suggest avoiding "we" (per IBM Style Guide).

Also:

```suggestion
NOTE: We now have the image. The next step is to release the image for which we need to set up the workspace.
```

Also, does this need to be a `NOTE`? If following CCS mod docs, there is the `.Next steps` list available, if you like.

== Setting up workspaces to release an FBC application

Two teams work together to release an application:

> * *Development team*- The team that develops and supports the application in a pre-production environment.

Reviewer 3: ```suggestion
* *Development team* - The team that develops and supports the application in a pre-production environment.
```

* *Managed environment team* - The team that supports the production instance of that application.

To release an application to a managed environment, the development team creates a `ReleasePlan` object in the developer workspace. The `ReleasePlan` object includes a reference to the application that the development team wants to release, along with the namespace and workspace where the application is supposed to release.

=== Creating a ReleasePlan object

.Prerequisites

> * You have an existing Development and Managed workspace.

Reviewer 3: This looks like the only place where "development" and "managed" are init-capped in reference to the workspaces. Suggest syncing up on either always-lowercase or always-init-cap:
```suggestion
* You have an existing development and managed workspace.
```

* You have installed `oc`.
* You have completed the steps listed in the link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/getting-started/getting_started_in_cli/[Getting started in the CLI] page. 

.Procedures

> * In the development workspace, create a `release_plan.yaml` object by running: `oc apply -f releaseplan.yaml -n dev`. It add the required configuration to your workspace that is needed to execute a release pipeline.

Reviewer 2: ```suggestion
* In the development workspace, create a `release_plan.yaml` object by running: `oc apply -f releaseplan.yaml -n dev`. It adds the required configuration to your workspace that is needed to execute a release pipeline in a managed workspace.
```

> * You have installed `oc`.

Reviewer 3: If you'd like to follow the `openshift-docs` [term glossary](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#openshift-cli) on this:
```suggestion
* You have installed the OpenShift CLI (`oc`).
```

> * In the development workspace, create a `release_plan.yaml` object by running: `oc apply -f releaseplan.yaml -n dev`. It adds the required configuration to your workspace that is needed to execute a release pipeline in a managed workspace.

Reviewer 2: You probably want to mention that there must be a corresponding releaseplanadmission in the managed workspace. If just one or the other exists, an application isn't configured for releases yet.

> * In the development workspace, create a `release_plan.yaml` object by running: `oc apply -f releaseplan.yaml -n dev`. It adds the required configuration to your workspace that is needed to execute a release pipeline in a managed workspace.

Reviewer 3: Can the command become a block so that it's not in-line? E.g.:
```suggestion
* In the development workspace, create a `release_plan.yaml` file by running the following command:
+
[source,terminal]
----
$ oc apply -f releaseplan.yaml -n dev
----
+
This command adds the required configuration to your workspace that is needed to execute a release pipeline in a managed workspace.
```

+
[source,yaml]
----
apiVersion: appstudio.redhat.com/v1alpha1
kind: ReleasePlan <1>
metadata:
  labels:
    release.appstudio.openshift.io/auto-release: 'true'
    release.appstudio.openshift.io/standing-attribution: 'true'
name: manual-fbc-release <2>
spec:
  application: fbc-example <3>
  displayName: manual-fbc-release 
  target: managed-release-team-tenant <4>

> ----

Reviewer 3: Is this the file that they are expected to create locally before running `oc apply` (e.g., the aforementioned `releaseplan.yaml`), or is this an example of something that changes _after_ running the `oc apply` command?

If the former, suggest maybe having this be 2 steps. First step: instruct to create the local `release_plan.yaml` file with the provided example code block, and you could add a title to the block like:
```
.Example `release_plan.yaml` file
[source,yaml]
----
```

Second step: run `oc apply` on the file.

<1> The name of the resource that you are creating.
<2> The name of the release plan.
<3> The name of the application that you want to deploy to the managed workspace.

> <4> The workspace to which the system deploys the application. This workspace is created by the Managed environment team (for example, your organization’s SRE team)

Reviewer 3: Should "Managed" by init-capped here? Other instances of "development team", for example, were lowercase. Also, puncutation:
```suggestion
<4> The workspace to which the system deploys the application. This workspace is created by the Managed environment team (for example, your organization’s SRE team).
```

.Verification

> . Browse the application you are creating and select the **Deployements** tab. The **Deployment** tab lists all the `ReleasePlan` objects that you created.

Reviewer 2: ```suggestion
. Browse to the application you are creating and select the **Deployments** tab. The **Deployment** tab lists all the `ReleasePlan` objects that you created.
```

Reviewer 2: The second sentence here isn't correct. The deployments tab shows how many instances of their app are running on static environments in their namespace. https://issues.redhat.com/browse/RHTAP-164 will add support for viewing and creating ReleasePlans in the UI. The Releases tab lists all attempted releases of their application to the managed workspace, once the releaseplan and corresponding releaseplanadmission are configured and matching.

If they want to validate that a release was successful, they should look at the Releases tab. There's no way to verify in the UI today that the releaseplan was created and correctly points to a releaseplanadmission.

. Review the appropriate `ReleasePlan` object for more information about the `ReleasePlan` objects you created. 

== Additional information

> For information about releasing an Application to production, see link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/proc_release_application/[Releasing an application].

Reviewer 3: ?
```suggestion
For information about releasing an application to production, see link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/proc_release_application/[Releasing an application].
```

:_content-type: PROCEDURE
:myfunctionone: maintaining_operator_graphs

[id="maintaining_operator_graphs_{context}"]
= Maintaining an Operator graph

{ProductName} uses File-based catalogs instead of SQLite or ClusterServiceVersion (CVS) to produce an Operator catalog. File-based catalogs (FBC), often referred to as “Operator catalogs”, provide a simpler way to store and share metadata about Operator graphs written for Operator Lifecycle Manager (OLM). With {ProductName}, you can easily build, test, and release multiple components from a single Operator catalog repository.

> {ProductName} only supports use of Operator catalogs to store the Operator graphs because of the following benefits and because they have been supported since OCP 4.11:

Reviewer 3: This sentence and the following bullets seem to be about highlighting the differences/benefits that FBCs specifically provide (and the specific FBC-only support by App Studio) so I wouldn't go with the generic "Operator catalog" for this part so much. Also, this is now the first-instance of "OCP", so it should now be spelled out/abbreviated here (and not in the later, previously-discussed spot).

```suggestion
{ProductName} only supports use of FBCs to store the Operator graphs because of the following benefits and because they have been supported since OpenShift Container Platform (OCP) 4.11:
```

* *Centralized graph updates:* Operator catalogs consolidates graph updates into a single catalog file, eliminating the need for iterative updates and simplifying the management of complex graphs.

* *Convenient management and flexibility:* Compared to SQLite, Operator catalogs offers a more convenient solution for managing graph updates. Users can define all updates in one place using file-based catalogs, streamlining the process and improving overall efficiency. This centralized management capability simplifies the process and improves overall efficiency.

* *Improved ownership and analysis:* Operator catalogs provides users with improved ownership and control over their graphs through clean and declarative templates for comprehensive analysis and surface level checks. This simplifies upgrade possibilities determination and error state recovery, eliminating the need to rely on limited configuration options found in localized CSV files.

For additional information, see link:https://olm.operatorframework.io/docs/reference/file-based-catalogs/[File-based Catalogs].

> == Setting up an FBC-compatible git repository

Reviewer 3: ```suggestion
== Setting up an FBC-compatible Git repository
```

> For additional information, see link:https://docs.openshift.com/container-platform/4.13/operators/understanding/olm-packaging-format.html#olm-file-based-catalogs_olm-packaging-format[File based Catalogs].

Reviewer: Possible outline for changing this flow --

* Reword introduction to having a supported flow for producing an Operator catalog which defines the upgrade graph for an Operator
* Acknowledge that two different strategies exist, SQLite and FBC, and the difference between them
* State that AppStudio only supports the latter because of its benefits and since it has been supported since OCP 4.11.

@adellape, fyi

> For additional information, see link:https://docs.openshift.com/container-platform/4.13/operators/understanding/olm-packaging-format.html#olm-file-based-catalogs_olm-packaging-format[File based Catalogs].

Reviewer 3: @gtrivedi88 Thanks for considering my earlier suggestion about using the downstream doc, however looking at this again, I see that it's using some newer schemas like `olm.composite.catalogs` which we haven't downstreamed for docs yet, so might actually be better to stick with the upstream link you had. Sorry about that!

. Returning to the Manage build pipelines page in {ProductName}, select *Send pull request*.
. Select *Merge in GitHub*.
. In GitHub, merge the pull request from the `red-hat-trusted-app-pipeline` bot.
. Let {ProductName} complete PipelineRun for the upgraded build pipeline.

.Verification

. Go to *Activity > Pipeline* runs.

. Select the most recent *PipelineRun*.

. View the `fbc-validate` and the `fbc-related-image-check` task within the build pipeline. Ensure that the build process is executing these tasks as part of the build process and that they are not indicating a “skip checks” status. 

+
If these tasks display a green checkmark, it indicates that the build process has successfully validated the FBC components against an existing link:https://enterprisecontract.dev/docs/ec-policies/release_policy.html#labels_package[Enterprise Contract] policy. However, if they do not show a green checkmark, it indicates some potential issues with the FBC components.

+
NOTE: The `fbc-validate` task checks the existence and functionality of binaries, and runs validation on configurations within an FBC image. 

.Next steps

> The application image is now available, and the next step is to release it by setting up the wrokspace.

Reviewer 3: ```suggestion
The application image is now available, and the next step is to release it by setting up the workspace.
```

== Setting up workspaces to release an FBC application

Two teams work together to release an application:

* *Development team* - The team that develops and supports the application in a pre-production environment.
* *Managed environment team* - The team that supports the production instance of that application.

To release an application to a managed environment, the development team creates a `ReleasePlan` object in the developer workspace. The `ReleasePlan` object includes a reference to the application that the development team wants to release, along with the namespace and workspace where the application is supposed to release.

=== Creating a ReleasePlan object

.Prerequisites

* You have an existing development and managed workspace.
* You have installed the OpenShift CLI (`oc`).
* You have completed the steps listed in the link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/getting-started/getting_started_in_cli/[Getting started in the CLI] page. 

.Procedures

> . Create a `release_plan.yaml` file locally.
> 
> +
> Example `release_plan.yaml` file:
> 
> +
> [source,yaml]

Reviewer 3: If the `.Title` style for blocks works like normal AsciiDoc on your site, then suggest the following (note change to `:` for lead-in sentence in this case, too):

```suggestion
. Create a `release_plan.yaml` file locally:
+
.Example `release_plan.yaml` file
[source,yaml]
```
